File: internal/server/groups.go
// SPDX-License-Identifier: Apache-2.0

package server

import (
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/go-chi/chi/v5"

	"github.com/gopherium/gouncer/authkit"

	"github.com/gopherium/gophenberg/internal/content"
)

// groupResponse is a field group as the admin API answers it.
type groupResponse struct {
	ID        int             `json:"id"`
	Title     string          `json:"title"`
	Location  content.Rules   `json:"location"`
	Position  int             `json:"position"`
	Active    bool            `json:"active"`
	Fields    []fieldResponse `json:"fields"`
	CreatedAt time.Time       `json:"created_at"`
	UpdatedAt time.Time       `json:"updated_at"`
}

// groupListResponse is the stored field groups as the admin API answers them.
type groupListResponse struct {
	Items []groupResponse `json:"items"`
}

// paramResponse is one rule source as the admin API answers it.
type paramResponse struct {
	Source    string           `json:"source"`
	Operators []string         `json:"operators"`
	Values    []choiceResponse `json:"values"`
}

// choiceResponse is one value a rule source offers.
type choiceResponse struct {
	Value string `json:"value"`
	Label string `json:"label"`
}

// paramListResponse is the rule sources a location may read.
type paramListResponse struct {
	Items []paramResponse `json:"items"`
}

// newGroupResponse builds a groupResponse, normalizing timestamps to UTC.
func newGroupResponse(g content.Group) groupResponse {
	fields := make([]fieldResponse, len(g.Fields))
	for i, f := range g.Fields {
		fields[i] = newFieldResponse(f)
	}
	location := g.Location
	if location == nil {
		location = content.Rules{}
	}
	return groupResponse{
		ID:        g.ID,
		Title:     g.Title,
		Location:  location,
		Position:  g.Position,
		Active:    g.Active,
		Fields:    fields,
		CreatedAt: g.CreatedAt.UTC(),
		UpdatedAt: g.UpdatedAt.UTC(),
	}
}

// newGroupListResponse builds the listing of every stored group.
func newGroupListResponse(groups []content.Group) groupListResponse {
	items := make([]groupResponse, len(groups))
	for i, g := range groups {
		items[i] = newGroupResponse(g)
	}
	return groupListResponse{Items: items}
}

// handleGroupList returns an http.HandlerFunc listing the stored field groups.
func (s *server) handleGroupList() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		groups, err := s.types.Groups(r.Context())
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusOK, newGroupListResponse(groups))
	}
}

// handleGroupParams returns an http.HandlerFunc listing the sources a location rule may read.
func (s *server) handleGroupParams() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		params := s.types.Params(r.Context()).All()
		items := make([]paramResponse, len(params))
		for i, p := range params {
			choices, err := p.Values(r.Context())
			if err != nil {
				respondDomainError(w, err)
				return
			}
			items[i] = paramResponse{
				Source: p.Name(), Operators: p.Operators(), Values: offeredChoices(choices),
			}
		}
		authkit.Respond(w, http.StatusOK, paramListResponse{Items: items})
	}
}

// offeredChoices returns the values a rule source offers as the admin reads them.
func offeredChoices(choices []content.Choice) []choiceResponse {
	offered := make([]choiceResponse, len(choices))
	for i, c := range choices {
		offered[i] = choiceResponse{Value: c.Value, Label: c.Label}
	}
	return offered
}

// handleGroupCreate returns an http.HandlerFunc storing a new field group.
func (s *server) handleGroupCreate() http.HandlerFunc {
	type request struct {
		Title    string        `json:"title"`
		Location content.Rules `json:"location"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		created, err := s.types.CreateGroup(r.Context(), content.Group{
			Title: req.Title, Location: req.Location,
		})
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusCreated, newGroupResponse(created))
	}
}

// handleGroupPatch returns an http.HandlerFunc carrying a group's title, location and resting flag.
func (s *server) handleGroupPatch() http.HandlerFunc {
	type request struct {
		Title    *string        `json:"title"`
		Location *content.Rules `json:"location"`
		Active   *bool          `json:"active"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		stored, err := s.storedGroup(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		if req.Title != nil {
			stored.Title = *req.Title
		}
		if req.Location != nil {
			stored.Location = *req.Location
		}
		if req.Active != nil {
			stored.Active = *req.Active
		}
		updated, err := s.types.UpdateGroup(r.Context(), stored)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusOK, newGroupResponse(updated))
	}
}

// handleGroupDelete returns an http.HandlerFunc removing a group with its fields and their values.
func (s *server) handleGroupDelete() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		if err := s.types.DeleteGroup(r.Context(), id); err != nil {
			respondDomainError(w, err)
			return
		}
		w.WriteHeader(http.StatusNoContent)
	}
}

// handleGroupOrder returns an http.HandlerFunc storing the order the groups are read in.
func (s *server) handleGroupOrder() http.HandlerFunc {
	type request struct {
		Order []int `json:"order"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		reordered, err := s.types.ReorderGroups(r.Context(), req.Order)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusOK, newGroupListResponse(reordered))
	}
}

// fieldCreateRequest is a field declaration as the admin API reads it.
type fieldCreateRequest struct {
	Key       string         `json:"key"`
	Label     string         `json:"label"`
	Kind      string         `json:"kind"`
	RelatesTo string         `json:"relates_to"`
	Many      bool           `json:"many"`
	Required  bool           `json:"required"`
	Settings  map[string]any `json:"settings"`
}

// handleGroupFieldCreate returns an http.HandlerFunc declaring a field inside a group.
func (s *server) handleGroupFieldCreate() http.HandlerFunc {
	type request = fieldCreateRequest
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		asked, err := content.NewField(content.Field{
			Key:       req.Key,
			Label:     req.Label,
			Kind:      content.FieldKind(req.Kind),
			RelatesTo: req.RelatesTo,
			Many:      req.Many,
			Required:  req.Required,
			Settings:  req.Settings,
		})
		if err != nil {
			respondDomainError(w, err)
			return
		}
		created, err := s.types.CreateFieldInGroup(r.Context(), id, asked)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusCreated, newFieldResponse(created))
	}
}

// handleSubFieldCreate returns an http.HandlerFunc declaring a field inside a container.
func (s *server) handleSubFieldCreate() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[fieldCreateRequest](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		parent, err := s.fieldAtPath(r, id)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		asked, err := content.NewSubField(content.Field{
			Key:       req.Key,
			Label:     req.Label,
			Kind:      content.FieldKind(req.Kind),
			RelatesTo: req.RelatesTo,
			Many:      req.Many,
			Required:  req.Required,
			Settings:  req.Settings,
		}, parent.Kind)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		created, err := s.types.CreateSubField(r.Context(), parent.ID, asked)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusCreated, newFieldResponse(created))
	}
}

// handleSubFieldDelete returns an http.HandlerFunc removing a field inside a container.
func (s *server) handleSubFieldDelete() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		held, err := s.fieldAtPath(r, id)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		if err := s.types.DeleteSubField(r.Context(), held.ID); err != nil {
			respondDomainError(w, err)
			return
		}
		w.WriteHeader(http.StatusNoContent)
	}
}

// fieldAtPath returns the field the request path addresses inside its group, however deep it stands.
func (s *server) fieldAtPath(r *http.Request, groupID int) (content.Field, error) {
	groups, err := s.types.Groups(r.Context())
	if err != nil {
		return content.Field{}, err
	}
	for _, g := range groups {
		if g.ID != groupID {
			continue
		}
		return fieldDown(g.Fields, strings.Split(chi.URLParam(r, "fieldPath"), "."))
	}
	return content.Field{}, content.ErrGroupNotFound
}

// fieldDown returns the field the keys address among the declared ones.
func fieldDown(declared []content.Field, keys []string) (content.Field, error) {
	for _, f := range declared {
		if f.Key != keys[0] {
			continue
		}
		if len(keys) == 1 {
			return f, nil
		}
		return fieldDown(f.Fields, keys[1:])
	}
	return content.Field{}, content.ErrFieldNotFound
}

// fieldPatchRequest is a field edit as the admin API reads it.
type fieldPatchRequest struct {
	Label     *string         `json:"label"`
	Required  *bool           `json:"required"`
	Settings  *map[string]any `json:"settings"`
	UpdatedAt *time.Time      `json:"updated_at"`
}

// handleGroupFieldPatch returns an http.HandlerFunc carrying a field's label, required flag and settings.
func (s *server) handleGroupFieldPatch() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[fieldPatchRequest](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		if req.UpdatedAt == nil {
			authkit.RespondError(w, http.StatusBadRequest, authkit.ErrorResponse{
				Message: "missing updated_at", Code: "body_field_required", Meta: map[string]any{"field": "updated_at"},
			})
			return
		}
		updated, err := s.patchGroupField(r, id, req)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusOK, newFieldResponse(updated))
	}
}

// patchGroupField applies the edit to the stored field once the expectation holds.
func (s *server) patchGroupField(r *http.Request, id int, req fieldPatchRequest) (content.Field, error) {
	keys := strings.Split(chi.URLParam(r, "fieldKey"), ".")
	stored, err := s.groupFieldDown(r, id, keys)
	if err != nil {
		return content.Field{}, err
	}
	if !req.UpdatedAt.Equal(stored.UpdatedAt) {
		return content.Field{}, content.ErrConflict
	}
	if req.Label != nil {
		stored.Label = *req.Label
	}
	if req.Required != nil {
		stored.Required = *req.Required
	}
	if req.Settings != nil {
		stored.Settings = *req.Settings
	}
	if len(keys) == 1 {
		return s.types.UpdateFieldInGroup(r.Context(), id, stored, *req.UpdatedAt)
	}
	return s.types.UpdateSubField(r.Context(), stored.ID, stored, *req.UpdatedAt)
}

// groupFieldDown returns the field the keys address inside the group, however deep it stands.
func (s *server) groupFieldDown(r *http.Request, groupID int, keys []string) (content.Field, error) {
	groups, err := s.types.Groups(r.Context())
	if err != nil {
		return content.Field{}, err
	}
	for _, g := range groups {
		if g.ID != groupID {
			continue
		}
		return fieldDown(g.Fields, keys)
	}
	return content.Field{}, content.ErrGroupNotFound
}

// handleSubFieldOrder returns an http.HandlerFunc standing the fields inside a container as asked.
func (s *server) handleSubFieldOrder() http.HandlerFunc {
	type request struct {
		Order []string `json:"order"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		held, err := s.fieldAtPath(r, id)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		reordered, err := s.types.ReorderSubFields(r.Context(), held.ID, req.Order)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		items := make([]fieldResponse, len(reordered))
		for i, f := range reordered {
			items[i] = newFieldResponse(f)
		}
		authkit.Respond(w, http.StatusOK, fieldListResponse{Items: items})
	}
}

// handleGroupFieldDelete returns an http.HandlerFunc removing a field and the values it held.
func (s *server) handleGroupFieldDelete() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		if err := s.types.DeleteFieldInGroup(r.Context(), id, chi.URLParam(r, "fieldKey")); err != nil {
			respondDomainError(w, err)
			return
		}
		w.WriteHeader(http.StatusNoContent)
	}
}

// handleGroupFieldOrder returns an http.HandlerFunc storing a group's field declaration order.
func (s *server) handleGroupFieldOrder() http.HandlerFunc {
	type request struct {
		Order []string `json:"order"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		reordered, err := s.types.ReorderFieldsInGroup(r.Context(), id, req.Order)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		items := make([]fieldResponse, len(reordered))
		for i, f := range reordered {
			items[i] = newFieldResponse(f)
		}
		authkit.Respond(w, http.StatusOK, fieldListResponse{Items: items})
	}
}

// handleGroupFieldMove returns an http.HandlerFunc carrying a field into another group.
func (s *server) handleGroupFieldMove() http.HandlerFunc {
	type request struct {
		ToGroup int `json:"to_group"`
	}
	return func(w http.ResponseWriter, r *http.Request) {
		id, err := groupIDOf(r)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		req, err := decodeKnown[request](w, r)
		if err != nil {
			respondBodyError(w, err)
			return
		}
		moved, err := s.types.MoveField(r.Context(), id, chi.URLParam(r, "fieldKey"), req.ToGroup)
		if err != nil {
			respondDomainError(w, err)
			return
		}
		authkit.Respond(w, http.StatusOK, newFieldResponse(moved))
	}
}

// storedGroup returns the group the request names, or the reason it holds none.
func (s *server) storedGroup(r *http.Request) (content.Group, error) {
	id, err := groupIDOf(r)
	if err != nil {
		return content.Group{}, err
	}
	groups, err := s.types.Groups(r.Context())
	if err != nil {
		return content.Group{}, err
	}
	for _, g := range groups {
		if g.ID == id {
			return g, nil
		}
	}
	return content.Group{}, content.ErrGroupNotFound
}

// groupIDOf returns the group identifier the path names, or the reason it names none.
func groupIDOf(r *http.Request) (int, error) {
	id, err := strconv.Atoi(chi.URLParam(r, "id"))
	if err != nil {
		return 0, content.Refuse(content.ErrGroupNotFound, "group_id_malformed",
			"content: the group identifier is not a number", nil)
	}
	return id, nil
}

File: frontend/src/content/groups.ts
// SPDX-License-Identifier: Apache-2.0

import { z } from 'zod'
import { __ } from '@wordpress/i18n'

import { errorText } from '../i18n/errors'
import { fieldSchema, toField } from './types'
import type { ContentField, NewField } from './types'

const ruleSchema = z.object({
	source: z.string(),
	operator: z.string(),
	value: z.string(),
})

const groupSchema = z.object({
	id: z.number(),
	key: z.string(),
	title: z.string(),
	location: z.array(z.array(ruleSchema)),
	position: z.number(),
	active: z.boolean(),
	fields: z.array(fieldSchema),
})

const groupListSchema = z.object({ items: z.array(groupSchema) })

const sourceSchema = z.object({
	source: z.string(),
	operators: z.array(z.string()),
	values: z.array(z.object({ value: z.string(), label: z.string() })),
})

const errorSchema = z.object({
	error: z.string(),
	code: z.string().optional(),
	meta: z.record(z.string(), z.unknown()).optional(),
})

/** One condition of a group's location, read by its source. */
export interface LocationRule {
	source: string
	operator: string
	value: string
}

/** Where a group appears, as OR groups of AND rules. */
export type Location = LocationRule[][]

/** A field group as the admin reads it. */
export interface FieldGroup {
	id: number
	title: string
	location: Location
	position: number
	active: boolean
	fields: ContentField[]
}

/** One source a location rule may read, with the choices it offers. */
export interface RuleSource {
	source: string
	operators: string[]
	values: { value: string; label: string }[]
}

/** The parts of a group an edit may move, where an absent part is unchanged. */
export interface GroupEdit {
	title?: string
	location?: Location
	active?: boolean
}

/**
 * Returns the admin view of a stored group.
 * @param row - The group as the API answered it.
 * @returns The group the admin reads.
 */
function toGroup(row: z.infer<typeof groupSchema>): FieldGroup {
	return {
		id: row.id,
		title: row.title,
		location: row.location,
		position: row.position,
		active: row.active,
		fields: row.fields.map(toField),
	}
}

/** The failure a write reports when the stored field moved on before it landed. */
export class StaleWriteError extends Error {}

/**
 * Throws the reason a group write was refused, or the status it failed with.
 * @param response - The answer the server gave.
 */
async function refuse(response: Response): Promise<never> {
	const parsed = errorSchema.safeParse(await response.json().catch(() => null))
	const said = errorText(parsed.success ? parsed.data : { error: '' })
	throw response.status === 409 ? new StaleWriteError(said) : new Error(said)
}

/**
 * Returns the sentence naming why a group write was turned away.
 * @param cause - What the write failed with.
 * @returns The sentence to show.
 */
export function groupErrorMessage(cause: unknown): string {
	return cause instanceof Error ? cause.message : __('The field groups could not be reached.', 'gophenberg')
}

/** The key the stored field groups are cached under. */
export const groupsQueryKey = ['field-groups']

/** The key the rule sources are cached under. */
export const ruleSourcesQueryKey = ['rule-sources']

/** The source a rule reads to match the content type. */
export const typeSource = 'content_type'

/** The value a rule carries to match every content type. */
export const anyType = '*'

/**
 * Returns every stored field group in position order.
 * @returns The groups with their rules and fields.
 */
export async function listGroups(): Promise<FieldGroup[]> {
	const response = await fetch('/api/groups')
	if (!response.ok) {
		throw new Error(`reading the field groups failed with status ${response.status}`)
	}
	return groupListSchema.parse(await response.json()).items.map(toGroup)
}

/**
 * Stores a new field group.
 * @param title - The name the group is listed under.
 * @param location - The rules deciding where its fields appear.
 * @returns The stored group.
 */
export async function createGroup(title: string, location: Location): Promise<FieldGroup> {
	const response = await fetch('/api/groups', {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({ title, location }),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toGroup(groupSchema.parse(await response.json()))
}

/**
 * Edits the parts of a group the caller names.
 * @param id - The group to edit.
 * @param edit - The parts to move.
 * @returns The stored group.
 */
export async function updateGroup(id: number, edit: GroupEdit): Promise<FieldGroup> {
	const body: Record<string, unknown> = {}
	if (edit.title !== undefined) {
		body.title = edit.title
	}
	if (edit.location !== undefined) {
		body.location = edit.location
	}
	if (edit.active !== undefined) {
		body.active = edit.active
	}
	const response = await fetch(`/api/groups/${id}`, {
		method: 'PATCH',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify(body),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toGroup(groupSchema.parse(await response.json()))
}

/**
 * Removes a group with its fields and the values they held.
 * @param id - The group to remove.
 */
export async function deleteGroup(id: number): Promise<void> {
	const response = await fetch(`/api/groups/${id}`, { method: 'DELETE' })
	if (!response.ok) {
		await refuse(response)
	}
}

/**
 * Stores the order the groups are read in.
 * @param ids - Every stored group in the order to store.
 * @returns The reordered groups.
 */
export async function reorderGroups(ids: number[]): Promise<FieldGroup[]> {
	const response = await fetch('/api/groups/order', {
		method: 'PUT',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({ order: ids }),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return groupListSchema.parse(await response.json()).items.map(toGroup)
}

/**
 * Returns the sources a location rule may read, with the choices each offers.
 * @returns The rule sources in registration order.
 */
export async function listRuleSources(): Promise<RuleSource[]> {
	const response = await fetch('/api/groups/params')
	if (!response.ok) {
		throw new Error(`reading the rule sources failed with status ${response.status}`)
	}
	return z.object({ items: z.array(sourceSchema) }).parse(await response.json()).items
}

/**
 * Declares a field inside a group.
 * @param id - The group declaring the field.
 * @param asked - The field to declare.
 * @returns The stored field.
 */
export async function createFieldInGroup(id: number, asked: NewField): Promise<ContentField> {
	const response = await fetch(`/api/groups/${id}/fields`, {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({
			key: asked.key,
			label: asked.label,
			kind: asked.kind,
			relates_to: asked.relatesTo,
			many: asked.many,
			required: asked.required,
			settings: asked.settings,
		}),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toField(fieldSchema.parse(await response.json()))
}

/**
 * Carries an edit to a field inside its group.
 * @param id - The group declaring the field.
 * @param key - The field to edit.
 * @param body - The parts to move.
 * @returns The stored field.
 */
async function patchFieldInGroup(
	id: number,
	key: string,
	body: Record<string, unknown>,
): Promise<ContentField> {
	const response = await fetch(`/api/groups/${id}/fields/${key}`, {
		method: 'PATCH',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify(body),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toField(fieldSchema.parse(await response.json()))
}

/**
 * Carries a new label for a field inside its group.
 * @param id - The group declaring the field.
 * @param key - The field to relabel.
 * @param label - The label to carry.
 * @param updatedAt - The timestamp the editor read.
 * @returns The stored field.
 */
export async function renameFieldInGroup(
	id: number,
	key: string,
	label: string,
	updatedAt: string,
): Promise<ContentField> {
	return patchFieldInGroup(id, key, { label, updated_at: updatedAt })
}

/**
 * Stores the settings a field carries, replacing the ones it held.
 * @param id - The group declaring the field.
 * @param key - The field to change.
 * @param settings - The settings to store.
 * @param updatedAt - The timestamp the editor read.
 * @returns The stored field.
 */
export async function setFieldSettingsInGroup(
	id: number,
	key: string,
	settings: Record<string, unknown>,
	updatedAt: string,
): Promise<ContentField> {
	return patchFieldInGroup(id, key, { settings, updated_at: updatedAt })
}

/**
 * Stores whether a field must be filled before its item publishes.
 * @param id - The group declaring the field.
 * @param key - The field to change.
 * @param required - Whether the field gates publishing.
 * @param updatedAt - The timestamp the editor read.
 * @returns The stored field.
 */
export async function setFieldRequiredInGroup(
	id: number,
	key: string,
	required: boolean,
	updatedAt: string,
): Promise<ContentField> {
	return patchFieldInGroup(id, key, { required, updated_at: updatedAt })
}

/**
 * Stores the declaration order of the fields a container holds.
 * @param id - The group declaring the container.
 * @param path - The dotted path naming the container.
 * @param keys - Every field key the container holds, in the order to store.
 * @returns The reordered fields.
 */
export async function reorderSubFields(
	id: number,
	path: string,
	keys: string[],
): Promise<ContentField[]> {
	const response = await fetch(`/api/groups/${id}/inside/${path}/order`, {
		method: 'PUT',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({ order: keys }),
	})
	if (!response.ok) {
		await refuse(response)
	}
	const listed = z.object({ items: z.array(fieldSchema) }).parse(await response.json())
	return listed.items.map(toField)
}

/**
 * Stores the declaration order of a group's fields.
 * @param id - The group whose fields are reordered.
 * @param keys - Every declared field key in the order to store.
 * @returns The fields in the stored order.
 */
export async function reorderFieldsInGroup(id: number, keys: string[]): Promise<ContentField[]> {
	const response = await fetch(`/api/groups/${id}/fields/order`, {
		method: 'PUT',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({ order: keys }),
	})
	if (!response.ok) {
		await refuse(response)
	}
	const listed = z.object({ items: z.array(fieldSchema) }).parse(await response.json())
	return listed.items.map(toField)
}

/**
 * Removes a field from its group with every value it held.
 * @param id - The group declaring the field.
 * @param key - The field to remove.
 */
export async function deleteFieldInGroup(id: number, key: string): Promise<void> {
	const response = await fetch(`/api/groups/${id}/fields/${key}`, { method: 'DELETE' })
	if (!response.ok) {
		await refuse(response)
	}
}

/**
 * Declares a field inside the container the path addresses.
 * @param id - The group declaring the container.
 * @param path - The dotted keys addressing the container.
 * @param asked - The field to declare.
 * @returns The stored field.
 */
export async function createSubField(
	id: number,
	path: string,
	asked: NewField,
): Promise<ContentField> {
	const response = await fetch(`/api/groups/${id}/fields/${path}`, {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({
			key: asked.key,
			label: asked.label,
			kind: asked.kind,
			relates_to: asked.relatesTo,
			many: asked.many,
			required: asked.required,
			settings: asked.settings,
		}),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toField(fieldSchema.parse(await response.json()))
}

/**
 * Removes the field the path addresses inside its container, and the values every item held under it.
 * @param id - The group declaring the container.
 * @param path - The dotted keys addressing the field.
 */
export async function deleteSubField(id: number, path: string): Promise<void> {
	const response = await fetch(`/api/groups/${id}/inside/${path}`, { method: 'DELETE' })
	if (!response.ok) {
		await refuse(response)
	}
}

/**
 * Carries a field into another group, keeping the values it holds.
 * @param id - The group the field leaves.
 * @param key - The field to carry.
 * @param toGroup - The group the field lands in.
 * @returns The stored field.
 */
export async function moveField(id: number, key: string, toGroup: number): Promise<ContentField> {
	const response = await fetch(`/api/groups/${id}/fields/${key}/move`, {
		method: 'POST',
		headers: { 'content-type': 'application/json' },
		body: JSON.stringify({ to_group: toGroup }),
	})
	if (!response.ok) {
		await refuse(response)
	}
	return toField(fieldSchema.parse(await response.json()))
}
